tools/api_proto_breaking_change_detector/detector.py

```diff
@@ -66,96 +52,147 @@ def lock_file_changed(self) -> bool:
+            path_to_changed_dir {str} -- absolute path to a directory containing proto files in the after state
+            buf_path {str} -- path to the buf binary (default: "buf")
+            additional_args {List[str]} -- additional arguments passed into the buf binary invocations
+            config_file_loc {str} -- absolute path to buf.yaml configuration file (if not provided, uses default buf configuration)
+
+            If using lock file mode:
+                path_to_lock_file {str} -- absolute path to the lock file representing the initial state (must be provided if using lock file mode)
+            If using git ref mode:
+                git_ref {str} -- git reference to use for the initial state of the protos (typically a commit hash) (must be provided if using git mode)
+                git_path {str} -- absolute path to .git file for the repository of interest (must be provided if using git mode)
+        """
+        if path_to_lock_file and not Path(path_to_lock_file).is_file():
+            raise ValueError(f"path_to_lock_file {path_to_lock_file} is not a file path")
 
-            if initial_code != 0 or len(initial_out) > 0 or len(initial_err) > 0:
-                raise ChangeDetectorInitializeError(
-                    f"Unexpected error during init:\n\tExit Status Code: {initial_code}\n\tstdout: {initial_out}\n\t stderr: {initial_err}\n"
-                )
+        if not Path(path_to_changed_dir).is_dir():
+            raise ValueError(f"path_to_changed_dir {path_to_changed_dir} is not a valid directory")
 
-            with open(lock_location, "r") as f:
-                self._initial_lock = f.readlines()
+        if Path.cwd() not in Path(path_to_changed_dir).parents:
+            raise ValueError(
+                f"path_to_changed_dir {path_to_changed_dir} must be a subdirectory of the cwd ({ Path.cwd() })"
+            )
 
-            copyfile(self._path_to_after, target)
+        if bool(path_to_lock_file) == bool(git_ref):
+            raise ValueError("Expecting either a path to a lock file or a git ref, but not both")
+        if bool(git_ref) != bool(git_path):
+            raise ChangeDetectorError(
+                "If using git mode, expecting a git ref and a path to .git file")
 
-            final_code, final_out, final_err = run_command(
-                ' '.join([buf_path, f"breaking --against {lock_location}", *buf_args]))
-            final_out, final_err = ''.join(final_out), ''.join(final_err)
+        self._path_to_changed_dir = path_to_changed_dir
+        self._additional_args = additional_args
+        self._final_lock = None
+        self._buf_path = buf_path or "buf"
+        self._config_file_loc = config_file_loc
+        self._path_to_lock_file = path_to_lock_file
+        self._git_ref = git_ref
+        self._git_path = git_path
+
+        if self._path_to_lock_file:
+            locktype = Path(self._path_to_lock_file).suffix
+            self._is_lock_text_file = locktype == '.json'
```

Comment:
I'm actually of the other opinion: I don't trust that Git's merge routine will produce the same JSON file that Buf would. Given two branches, A and B, off of main, Git will auto-merge them if they don't touch the same files, or if the changes they make to a file can be merged because they touch disjoint regions of the file. So our CI checks may pass happily for PRs for A and B, but then fail for the new main' = main + A + B because the merged JSON doesn't match what Buf would produce. If we use binary lock files, we do require that any PR that touches the API protos merge main in before submission, but that seems safer than the bad merge. I'm happy to be overruled though, or to say let's use JSON until this becomes a real issue (which is unlikely since we're now using lock files only in tests, IIRC).

Reply:
I think it really depends on what buf puts in the lock file contents, and IIUC buf essentially serializes relevant proto-objects from all proto files into a single list, so assuming that changes are only restricted to a specific region, I think it might be ok.
In any case, it might be that we no longer need the lock file support (as we might be able to compare against HEAD also in tests).